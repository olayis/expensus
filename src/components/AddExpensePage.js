import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { startChangeCurrency } from '../actions/locale';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.startChangeCurrency({ currency: 'Danish Krone' });
    this.props.history.push('/');
  };
  render() {
    return (
      <main className='main-content'>
        <div className='content-container'>
          <div className='add-expense-container'>
            <ExpenseForm onSubmit={this.onSubmit} />
          </div>
        </div>
      </main>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
  startChangeCurrency: (currency) => dispatch(startChangeCurrency(currency)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
