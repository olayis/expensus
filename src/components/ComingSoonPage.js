import React from 'react';
import { Link } from 'react-router-dom';
import programming from '../../public/img/programming.svg';

const ComingSoonPage = () => (
  <main className='main-content'>
    <div className='content-container'>
      <div className='error'>
        <img
          src={programming}
          alt='Programmer'
          className='error__illustration'
          width='1075'
          height='585'
        />
        <h2 className='error__text'>This page is coming soon.</h2>
        <div className='error__explanation'>
          Continue managing your expenses.
          <span className='error__action'>
            <Link className='btn btn--try-again' to='/'>
              Go Home
            </Link>
          </span>
        </div>
      </div>
    </div>
  </main>
);

export default ComingSoonPage;
