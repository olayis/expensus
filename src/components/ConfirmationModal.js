import React from 'react';
import Modal from 'react-modal';

const ConfirmationModal = (props) => {
  if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#app');

  return (
    <div>
      <Modal
        contentLabel='Are you sure?'
        isOpen={!!props.deleteSelected}
        onRequestClose={props.stopDeleteDataSelected}
        closeTimeoutMS={200}
        className='modal'
        aria={{
          labelledby: 'heading',
          describedby: 'full_description',
        }}
      >
        <h3 id='heading' className='modal__title'>
          Delete Data
        </h3>
        <p id='full_description' className='modal__body'>
          Are you sure you want to delete all of your data on Expensus?
        </p>
        <button className='modal__btn--delete' onClick={props.handleDeleteData}>
          Delete
        </button>
        <button
          className='modal__btn--cancel'
          onClick={props.stopDeleteDataSelected}
        >
          Cancel
        </button>
      </Modal>
    </div>
  );
};

export default ConfirmationModal;
