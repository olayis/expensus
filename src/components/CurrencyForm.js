import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faExclamationCircle } from '@fortawesome/free-solid-svg-icons';

export default class CurrencyForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currency: props.currency,
      currencyError: '',
    };
  }

  onCurrencyChange = (e) => {
    const currency = e.target.value;
    this.setState(() => ({ currency }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.currency) {
      this.setState(() => ({
        currencyError: 'Please select a valid currency',
      }));
    } else {
      this.setState(() => ({ currencyError: '' }));
      this.props.onSubmit({
        currency: this.state.currency.trim(),
      });
    }
  };

  render() {
    return (
      <form className='input-field' onSubmit={this.onSubmit}>
        <div className='input-field__wrapper'>
          <select
            className='dropdown dropdown--currency'
            value={this.state.currency}
            onChange={this.onCurrencyChange}
            aria-label='Select Currency'
            name='currency'
          >
            <option value=''>-- Select Currency --</option>
            <option value='Baht'>Baht (฿)</option>
            <option value='Danish Krone'>Danish Krone (DKK)</option>
            <option value='Dollar'>Dollar ($)</option>
            <option value='Dong'>Dong (₫)</option>
            <option value='Euro'>Euro (€)</option>
            <option value='Forint'>Forint (Ft)</option>
            <option value='Hryvnia'>Hryvnia (₴)</option>
            <option value='Koruna'>Koruna (Kč)</option>
            <option value='Lira'>Lira (₺)</option>
            <option value='Lev'>Lev (лв)</option>
            <option value='Naira'>Naira (₦)</option>
            <option value='Norwegian krone'>Norwegian krone (kr)</option>
            <option value='Pound Sterling'>Pound Sterling (£)</option>
            <option value='Rand'>Rand (R)</option>
            <option value='Real'>Real (R$)</option>
            <option value='Ruble'>Ruble (₽)</option>
            <option value='Swiss Franc'>Swiss Franc (CHF)</option>
            <option value='Yen'>Yen (¥)</option>
            <option value='Yuan'>Yuan (¥)</option>
            <option value='Zloty'>Złoty (PLN)</option>
          </select>
          {this.state.currencyError && (
            <span className='input-field__error-msg'>
              <FontAwesomeIcon className='fa-icon' icon={faExclamationCircle} />
              {this.state.currencyError}
            </span>
          )}
        </div>
        <button className='btn btn-currency'>Change Currency</button>
      </form>
    );
  }
}
