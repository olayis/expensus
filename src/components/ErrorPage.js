import React from 'react';
import { history } from '../routers/AppRouter';
import engineer from '../../public/img/QA-Engineer.svg';

const ErrorPage = () => (
  <div className='content-container'>
    <div className='error'>
      <img
        src={engineer}
        alt='A Quality Assurance Engineer'
        className='error__illustration'
        width='1075'
        height='585'
      />
      <h2 className='error__text'>Oops! Something went wrong.</h2>
      <div className='error__explanation'>
        Click here to try again.
        <span className='error__action'>
          <button className='btn btn--try-again' onClick={() => history.go(0)}>
            Try Again
          </button>
        </span>
      </div>
    </div>
  </div>
);

export default ErrorPage;
