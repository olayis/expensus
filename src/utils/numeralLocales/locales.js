// Acknowledgement: numeral locale configurations were gotten from https://github.com/adamwdraper/Numeral-js

import numeral from 'numeral';

export default (locale) => {
  switch (locale) {
    case 'Baht':
      // Thai Baht
      numeral.register('locale', 'th', {
        delimiters: {
          thousands: ',',
          decimal: '.',
        },
        abbreviations: {
          thousand: 'พัน',
          million: 'ล้าน',
          billion: 'พันล้าน',
          trillion: 'ล้านล้าน',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: '฿',
        },
      });
      return numeral.locale('th');

    case 'Danish Krone':
      // Danish Denmark Krone
      numeral.register('locale', 'da-dk', {
        delimiters: {
          thousands: '.',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'k',
          million: 'mio',
          billion: 'mia',
          trillion: 'b',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: 'DKK',
        },
      });
      return numeral.locale('da-dk');

    case 'Dollar':
      return numeral.locale('en');

    case 'Euro':
      numeral.register('locale', 'de', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: '€',
        },
      });
      return numeral.locale('de');

    case 'Dong':
      // Vietnamese Dong
      numeral.register('locale', 'vi', {
        delimiters: {
          thousands: '.',
          decimal: ',',
        },
        abbreviations: {
          thousand: ' nghìn',
          million: ' triệu',
          billion: ' tỷ',
          trillion: ' nghìn tỷ',
        },
        ordinal: () => '.',
        currency: {
          symbol: '₫',
        },
      });
      return numeral.locale('vi');

    case 'Forint':
      // Hungarian Forint
      numeral.register('locale', 'hu', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'E', // ezer
          million: 'M', // millió
          billion: 'Mrd', // milliárd
          trillion: 'T', // trillió
        },
        ordinal: (number) => '.',
        currency: {
          symbol: ' Ft',
        },
      });
      return numeral.locale('hu');

    case 'Hryvnia':
      // Ukrainian Hryvnia
      numeral.register('locale', 'uk-ua', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'тис.',
          million: 'млн',
          billion: 'млрд',
          trillion: 'блн',
        },
        ordinal: () => {
          /* not ideal, but since in Ukrainian it can take on
           ** different forms (masculine, feminine, neuter)
           ** this is all we can do
           ** return '';
           */
        },
        currency: {
          symbol: '\u20B4',
        },
      });
      return numeral.locale('uk-ua');

    case 'Koruna':
      // Czech Koruna
      numeral.register('locale', 'cs', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'tis.',
          million: 'mil.',
          billion: 'b',
          trillion: 't',
        },
        ordinal: () => '.',
        currency: {
          symbol: 'Kč',
        },
      });
      return numeral.locale('cs');

    case 'Lira':
      // Turkish Lira
      const suffixes = {
        1: "'inci",
        5: "'inci",
        8: "'inci",
        70: "'inci",
        80: "'inci",

        2: "'nci",
        7: "'nci",
        20: "'nci",
        50: "'nci",

        3: "'üncü",
        4: "'üncü",
        100: "'üncü",

        6: "'ncı",

        9: "'uncu",
        10: "'uncu",
        30: "'uncu",

        60: "'ıncı",
        90: "'ıncı",
      };

      numeral.register('locale', 'tr', {
        delimiters: {
          thousands: '.',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'bin',
          million: 'milyon',
          billion: 'milyar',
          trillion: 'trilyon',
        },
        ordinal: (number) => {
          if (number === 0) {
            // special case for zero
            return "'ıncı";
          }

          const a = number % 10,
            b = (number % 100) - a,
            c = number >= 100 ? 100 : null;

          return suffixes[a] || suffixes[b] || suffixes[c];
        },
        currency: {
          symbol: '\u20BA',
        },
      });
      return numeral.locale('tr');

    case 'Lev':
      // Bulgarian Lev
      numeral.register('locale', 'bg', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          // I found these here http://www.unicode.org/cldr/charts/28/verify/numbers/bg.html
          thousand: 'хил',
          million: 'млн',
          billion: 'млрд',
          trillion: 'трлн',
        },
        ordinal: (number) => {
          // google translate suggests:
          // 1st=1-ви; 2nd=2-ри; 7th=7-ми;
          // 8th=8-ми and many others end with -ти
          // for example 3rd=3-ти
          // However since I've seen suggestions that in
          // Bulgarian the ordinal can be taken in
          // different forms (masculine, feminine, neuter)
          // I've opted to wimp out on commiting that to code
          return '';
        },
        currency: {
          symbol: 'лв',
        },
      });
      return numeral.locale('bg');

    case 'Naira':
      // Nigerian Naira
      numeral.register('locale', 'ng', {
        delimiters: {
          thousands: ',',
          decimal: '.',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: () => {
          const b = number % 10;
          return ~~((number % 100) / 10) === 1
            ? 'th'
            : b === 1
            ? 'st'
            : b === 2
            ? 'nd'
            : b === 3
            ? 'rd'
            : 'th';
        },
        currency: {
          symbol: '₦',
        },
      });
      return numeral.locale('ng');

    case 'Norwegian krone':
      numeral.register('locale', 'no', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: 'kr',
        },
      });
      return numeral.locale('no');

    case 'Pound Sterling':
      // locale: United Kingdom
      numeral.register('locale', 'en-gb', {
        delimiters: {
          thousands: ',',
          decimal: '.',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: (number) => {
          const b = number % 10;
          return ~~((number % 100) / 10) === 1
            ? 'th'
            : b === 1
            ? 'st'
            : b === 2
            ? 'nd'
            : b === 3
            ? 'rd'
            : 'th';
        },
        currency: {
          symbol: '£',
        },
      });
      return numeral.locale('en-gb');

    case 'Rand':
      // South Africa Rand
      numeral.register('locale', 'en-za', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: () => {
          const b = number % 10;
          return ~~((number % 100) / 10) === 1
            ? 'th'
            : b === 1
            ? 'st'
            : b === 2
            ? 'nd'
            : b === 3
            ? 'rd'
            : 'th';
        },
        currency: {
          symbol: 'R',
        },
      });
      return numeral.locale('en-za');

    case 'Real':
      // Brazilian Real
      numeral.register('locale', 'pt-br', {
        delimiters: {
          thousands: '.',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'mil',
          million: 'milhões',
          billion: 'b',
          trillion: 't',
        },
        ordinal: (number) => 'º',
        currency: {
          symbol: 'R$',
        },
      });
      return numeral.locale('pt-br');

    case 'Ruble':
      // Russian Ruble
      numeral.register('locale', 'ru', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'тыс.',
          million: 'млн.',
          billion: 'млрд.',
          trillion: 'трлн.',
        },
        ordinal: () => {
          // not ideal, but since in Russian it can taken on
          // different forms (masculine, feminine, neuter)
          // this is all we can do
          return '.';
        },
        currency: {
          symbol: '₽',
        },
      });
      return numeral.locale('ru');

    case 'Swiss Franc':
      // German in Switzerland
      numeral.register('locale', 'de-ch', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'k',
          million: 'm',
          billion: 'b',
          trillion: 't',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: 'CHF',
        },
      });
      return numeral.locale('de-ch');

    case 'Yen':
      // Japanese Yen
      numeral.register('locale', 'ja', {
        delimiters: {
          thousands: ',',
          decimal: '.',
        },
        abbreviations: {
          thousand: '千',
          million: '百万',
          billion: '十億',
          trillion: '兆',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: '¥',
        },
      });
      return numeral.locale('ja');

    case 'Yuan':
      // Simplified Chinese Yuan
      numeral.register('locale', 'chs', {
        delimiters: {
          thousands: ',',
          decimal: '.',
        },
        abbreviations: {
          thousand: '千',
          million: '百万',
          billion: '十亿',
          trillion: '兆',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: '¥',
        },
      });
      return numeral.locale('chs');

    case 'Zloty':
      // Polish Złoty
      numeral.register('locale', 'pl', {
        delimiters: {
          thousands: ' ',
          decimal: ',',
        },
        abbreviations: {
          thousand: 'tys.',
          million: 'mln',
          billion: 'mld',
          trillion: 'bln',
        },
        ordinal: (number) => '.',
        currency: {
          symbol: 'PLN',
        },
      });
      return numeral.locale('pl');

    default:
      return numeral.locale('en');
  }
};
