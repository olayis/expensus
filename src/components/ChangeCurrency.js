import React from 'react';
import { connect } from 'react-redux';
import { startChangeLocale } from '../actions/locale';
import CurrencyForm from './CurrencyForm';

export class ChangeCurrency extends React.Component {
  onSubmit = ({ currency }) => {
    this.props.startChangeLocale({ currency });
    this.props.history.push('/');
    this.props.history.go(0);
  };

  goBack = () => {
    this.props.history.goBack();
  };

  render() {
    return (
      <main className='main-content'>
        <div className='content-container'>
          <div className='component-container component-container--currency'>
            <CurrencyForm
              onSubmit={this.onSubmit}
              currency={this.props.currency}
              goBack={this.goBack}
            />
          </div>
        </div>
      </main>
    );
  }
}

const mapStateToProps = (state) => ({
  currency: state.locale ? state.locale.currency : '',
});

const mapDispatchToProps = (dispatch) => ({
  startChangeLocale: (currency) => dispatch(startChangeLocale(currency)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ChangeCurrency);
