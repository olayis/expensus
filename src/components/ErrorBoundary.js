import React from 'react';
import store from '../store/configureStore';
import ErrorPage from './ErrorPage';
import startLogError from '../actions/logError';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    process.env.NODE_ENV === 'Production'
      ? store.dispatch(startLogError(error, errorInfo))
      : '';
  }

  render() {
    if (this.state.hasError) {
      return <ErrorPage />;
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
