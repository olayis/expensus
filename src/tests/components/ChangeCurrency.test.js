import React from 'react';
import { shallow } from 'enzyme';
import { ChangeCurrency } from '../../components/ChangeCurrency';
import locales from '../fixtures/locales';

let startChangeLocale, history, wrapper;

beforeEach(() => {
  startChangeLocale = jest.fn();
  history = { push: jest.fn(), go: jest.fn() };
  wrapper = shallow(
    <ChangeCurrency startChangeLocale={startChangeLocale} history={history} />
  );
});

test('should render ChangeCurrency correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('CurrencyForm').prop('onSubmit')(locales[0]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startChangeLocale).toHaveBeenLastCalledWith(locales[0]);
});
