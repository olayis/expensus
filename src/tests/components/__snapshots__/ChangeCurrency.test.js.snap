// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ChangeCurrency correctly 1`] = `
<main
  className="main-content"
>
  <div
    className="content-container"
  >
    <div
      className="component-container component-container--currency"
    >
      <CurrencyForm
        goBack={[Function]}
        onSubmit={[Function]}
      />
    </div>
  </div>
</main>
`;
